astropy/units/core.py

```diff
@@ -2061,7 +2048,7 @@ def __call__(
                 )
 
         # now decide what we really need to do; define derived Unit?
-        if isinstance(represents, UnitBase):
+        if represents is not None:
```

Comment:
Inspired by your catch that `represents` does not need to be converted: it seems to me this clause really can be on top (before the `isinstance(s, Quantity)` - nothing but a string for `s` is going to work when `represents` is given.

Reply:
I was happy enough with the simplifications I had made that I hadn't bothered investigating further. But now that you pointed this out I had a look at `NamedUnit.__init__()` and that indeed does not handle `Quantity` instances.

Type checkers could have found this if more of our code was annotated.